librz/include/rz_il/rzil_opcodes.h

```diff
@@ -0,0 +1,447 @@
+// SPDX-FileCopyrightText: 2021 heersin <teablearcher@gmail.com>
+// SPDX-License-Identifier: LGPL-3.0-only
+
+#ifndef RZIL_OPCODES_H
+#define RZIL_OPCODES_H
+
+#include "definitions/definitions.h"
+
+/**
+ * \file rzil_opcodes.h
+ * \brief signatures of core theory opcodes
+ *
+ * Modular Hierarchy of the whole core theory
+ * (we implement the Minimal part only)
+ *                          o Core
+ *                            |
+ *             Trans o--------+--------o Float
+ *                            |        |
+ *                            o Basic  o FBasic
+ *                            |
+ *                            o Minimal
+ *                            |
+ *           +-------+--------+--------+-------+
+ *           |       |        |        |       |
+ *           o       o        o        o       o
+ *         Init    RzILBool     Bitv    Memory   Effect
+ *
+ * See also the references :
+ * 0. A gentle introduction to core theory http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/index.html
+ * 1. http://binaryanalysisplatform.github.io/bap/api/odoc/bap-core-theory/Bap_core_theory/Theory/index.html
+ * 2. For core and array theories https://smtlib.cs.uiowa.edu/theories.shtml
+ */
+
+/**
+ *  \struct rzil_op_int_t
+ *  \brief op structure for `int` (val int : 's Bitv.t Value.sort -> word -> 's bitv)
+ *
+ *  int s x is a bitvector constant x of sort s.
+ */
+struct rzil_op_int_t {
+	ut32 length; ///< s -- sort(type), length of bitvector
+	int value; ///< x -- value of bitvector
+};
+
+/**
+ *  \struct rzil_op_msb_lsb_t
+ *  \brief op structure for `msb` and `lsb` ('s bitv -> bool)
+ *  [MSB] msb x is the most significant bit of x.
+ *  [LSB] lsb x is the least significant bit of x.
+ */
+struct rzil_op_msb_lsb_t {
+	RzIL_BITV void *bv; ///< index of bitvector operand
```

Comment:
Why are all of these types only `void *`? It appears to me like they could be just `RzILOp *`.

On another note, have you considered splitting up `RzILOp` into multiple distinct structs depending on their return types?
So you would have `RzILOpBV`, `RzILOpBool` and `RzILOpEffect`, each with a union inside covering only the cases for the respective type.

Reply:
@thestr4ng3r  Oh the separating looks like a good idea, I will consider that. As for the first question, the reason to use `void*` is `RzILOp` defined in another file and relies on different `rzil_op_*_t`, and thus I cannot use `RzILOp` inside `rzil_op_*_t`. 

I used `typedef struct rzil_op_msb_lsb_t* RzILOpMsb` to bypass this limitation. Do you have better idea ?